modules/substrate/src/justification.rs

```diff
@@ -0,0 +1,329 @@
+	let ancestry_chain = AncestryChain::new(&justification.votes_ancestries);
+	match finality_grandpa::validate_commit(&justification.commit, &authorities_set, &ancestry_chain) {
+		Ok(ref result) if result.ghost().is_some() => {}
+		_ => return Err(Error::InvalidJustificationCommit),
+	}
+
+	// now that we know that the commit is correct, check authorities signatures
+	let mut buf = Vec::new();
+	let mut visited_hashes = BTreeSet::new();
+	for signed in &justification.commit.precommits {
+		if !sp_finality_grandpa::check_message_signature_with_buffer(
+			&finality_grandpa::Message::Precommit(signed.precommit.clone()),
+			&signed.id,
+			&signed.signature,
+			justification.round,
+			authorities_set_id,
+			&mut buf,
+		) {
+			return Err(Error::InvalidAuthoritySignature);
+		}
+
+		if justification.commit.target_hash == signed.precommit.target_hash {
+			continue;
+		}
+
+		match ancestry_chain.ancestry(justification.commit.target_hash, signed.precommit.target_hash) {
+			Ok(route) => {
+				// ancestry starts from parent hash but the precommit target hash has been visited
+				visited_hashes.insert(signed.precommit.target_hash);
+				visited_hashes.extend(route);
+			}
+			_ => {
+				// could this happen in practice? I don't think so, but original code has this check
+				return Err(Error::InvalidPrecommitAncestryProof);
+			}
+		}
+	}
+
+	let ancestry_hashes = justification
+		.votes_ancestries
+		.iter()
+		.map(|h: &Header| h.hash())
+		.collect();
+	if visited_hashes != ancestry_hashes {
+		return Err(Error::InvalidPrecommitAncestries);
+	}
+
+	Ok(())
+}
+
+/// GRANDPA justification of the bridged chain
+#[derive(Decode, RuntimeDebug)]
+#[cfg_attr(test, derive(codec::Encode))]
+struct GrandpaJustification<Header: HeaderT> {
+	round: u64,
+	commit: finality_grandpa::Commit<Header::Hash, Header::Number, AuthoritySignature, AuthorityId>,
+	votes_ancestries: Vec<Header>,
+}
+
+/// A utility trait implementing `finality_grandpa::Chain` using a given set of headers.
+struct AncestryChain<Header: HeaderT> {
+	ancestry: BTreeMap<Header::Hash, Header::Hash>,
+}
+
+impl<Header: HeaderT> AncestryChain<Header> {
+	fn new(ancestry: &[Header]) -> AncestryChain<Header> {
+		AncestryChain {
+			ancestry: ancestry
+				.iter()
+				.map(|header| (header.hash(), *header.parent_hash()))
+				.collect(),
+		}
+	}
+}
+
+impl<Header: HeaderT> finality_grandpa::Chain<Header::Hash, Header::Number> for AncestryChain<Header>
+where
+	Header::Number: finality_grandpa::BlockNumberOps,
+{
+	fn ancestry(&self, base: Header::Hash, block: Header::Hash) -> Result<Vec<Header::Hash>, GrandpaError> {
+		let mut route = Vec::new();
+		let mut current_hash = block;
+		loop {
+			if current_hash == base {
+				break;
+			}
+			match self.ancestry.get(&current_hash).cloned() {
+				Some(parent_hash) => {
+					current_hash = parent_hash;
+					route.push(current_hash);
+				}
+				_ => return Err(GrandpaError::NotDescendent),
+			}
+		}
+		route.pop(); // remove the base
```

Comment:
Hmm, why does the comment say "remove the base", if we never add the base to `route` since we exit the loop early in case `current_hash == base`? I guess this will remove `base.parent_hash`, no?

Reply:
If we exit loop early, then the vec would be empty && pop will be a noop, which, I believe is the expected result

Reply:
I was thinking about the non-empty case, but on a second look I got it.. The resulting vec will contain `[block.parent_hash, base]` after we exit the loop, because we go through the headers in reverse order (i.e. we push `next_block(base).parent_hash` as the last step in the loop, and then exit when `next_block(base).parent_hash == current_hash == base`, so we need to `pop` the base out of the vec.